File: workers-api/src/lib/ai.ts
import { BaziInput, BaziResult } from "./bazi";
import { GLM46Client, createGLM46Client } from "./glm46";

export interface AIAnalysisRequest {
  baziId: string;
  language?: "zh" | "en";
  aspects?: string[]; // 分析方面，如"career", "marriage", "health"等
  provider?: "deepseek" | "glm46"; // AI提供商选择
}

export interface AIAnalysisResult {
  analysis: string;
  aspects: string[];
  suggestions: string[];
  luckyPeriods: string[];
  warnings: string[];
  provider: string;
}

export async function analyzeBaziWithAI(
  baziInput: BaziInput,
  baziResult: BaziResult,
  language: "zh" | "en" = "zh",
  aspects?: string[],
  provider: "deepseek" | "glm46" = "deepseek",
): Promise<AIAnalysisResult> {
  try {
    let analysis: string;
    let providerName: string;

    if (provider === "glm46") {
      // 使用GLM-4.6进行分析
      const glm46Client = createGLM46Client();
      analysis = await glm46Client.analyzeBazi(
        baziInput,
        baziResult,
        language,
        aspects,
      );
      providerName = "GLM-4.6";
    } else {
      // 使用DeepSeek进行分析（原有逻辑）
      const prompt = buildPrompt(baziInput, baziResult, language, aspects);
      const systemPrompt = getSystemPrompt(language);

      const response = await fetch(
        "https://api.deepseek.com/v1/chat/completions",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${process.env.DEEPSEEK_API_KEY}`,
          },
          body: JSON.stringify({
            model: "deepseek-chat",
            messages: [
              { role: "system", content: systemPrompt },
              { role: "user", content: prompt },
            ],
            temperature: 0.7,
            max_tokens: 1500,
          }),
        },
      );

      if (!response.ok) {
        throw new Error(`DeepSeek API error: ${response.status}`);
      }

      const data = await response.json();
      analysis = data.choices[0]?.message?.content || "";
      providerName = "DeepSeek";
    }

    // 解析AI返回的内容
    const result = parseAIResponse(analysis, language, providerName);
    return result;
  } catch (error) {
    console.error(`AI analysis error (${provider}):`, error);

    // 如果指定了provider失败，尝试fallback到另一个provider
    if (provider === "glm46" && process.env.DEEPSEEK_API_KEY) {
      console.log("Falling back to DeepSeek...");
      return analyzeBaziWithAI(
        baziInput,
        baziResult,
        language,
        aspects,
        "deepseek",
      );
    } else if (provider === "deepseek" && process.env.GLM46_API_KEY) {
      console.log("Falling back to GLM-4.6...");
      return analyzeBaziWithAI(
        baziInput,
        baziResult,
        language,
        aspects,
        "glm46",
      );
    }

    throw new Error(
      `Failed to analyze bazi with ${provider}: ${error.message}`,
    );
  }
}

// 新增：仅使用GLM-4.6进行专门分析
export async function analyzeBaziWithGLM46(
  baziInput: BaziInput,
  baziResult: BaziResult,
  language: "zh" | "en" = "zh",
  aspects?: string[],
): Promise<AIAnalysisResult> {
  try {
    const glm46Client = createGLM46Client();
    const analysis = await glm46Client.analyzeBazi(
      baziInput,
      baziResult,
      language,
      aspects,
    );

    const result = parseAIResponse(analysis, language);
    result.provider = "GLM-4.6";
    return result;
  } catch (error) {
    console.error("GLM-4.6 analysis error:", error);
    throw new Error(`Failed to analyze bazi with GLM-4.6: ${error.message}`);
  }
}

// 新增：对比分析功能
export async function compareAIAnalysis(
  baziInput: BaziInput,
  baziResult: BaziResult,
  language: "zh" | "en" = "zh",
  aspects?: string[],
): Promise<{ deepseek: AIAnalysisResult; glm46: AIAnalysisResult }> {
  if (!process.env.DEEPSEEK_API_KEY || !process.env.GLM46_API_KEY) {
    throw new Error(
      "Both DEEPSEEK_API_KEY and GLM46_API_KEY are required for comparison",
    );
  }

  try {
    const [deepseekResult, glm46Result] = await Promise.all([
      analyzeBaziWithAI(baziInput, baziResult, language, aspects, "deepseek"),
      analyzeBaziWithAI(baziInput, baziResult, language, aspects, "glm46"),
    ]);

    return {
      deepseek: deepseekResult,
      glm46: glm46Result,
    };
  } catch (error) {
    console.error("AI comparison error:", error);
    throw new Error(`Failed to compare AI analysis: ${error.message}`);
  }
}

function buildPrompt(
  baziInput: BaziInput,
  baziResult: BaziResult,
  language: "zh" | "en",
  aspects?: string[],
): string {
  if (language === "en") {
    return `
    Please analyze the following Bazi (Four Pillars of Destiny) chart:

    Personal Information:
    - Name: ${baziInput.name || "Unknown"}
    - Gender: ${baziInput.gender}
    - Birth Date: ${baziInput.year}-${baziInput.month}-${baziInput.day} ${baziInput.hour}:${baziInput.minute}

    Bazi Chart:
    - Year Pillar: ${baziResult.yearPillar}
    - Month Pillar: ${baziResult.monthPillar}
    - Day Pillar: ${baziResult.dayPillar}
    - Hour Pillar: ${baziResult.hourPillar}
    - Day Master: ${baziResult.dayMaster}
    - Zodiac: ${baziResult.zodiac}
    - Nayin: ${baziResult.nayin}
    - Season: ${baziResult.season}

    Five Elements Distribution:
    - Wood: ${baziResult.fiveElements["木"]}
    - Fire: ${baziResult.fiveElements["火"]}
    - Earth: ${baziResult.fiveElements["土"]}
    - Metal: ${baziResult.fiveElements["金"]}
    - Water: ${baziResult.fiveElements["水"]}

    ${aspects ? `Please focus on these aspects: ${aspects.join(", ")}` : "Please provide a comprehensive analysis covering career, relationships, health, and wealth."}

    Please provide:
    1. A detailed personality analysis
    2. Strengths and weaknesses
    3. Career prospects
    4. Relationship compatibility
    5. Health considerations
    6. Wealth potential
    7. Lucky periods and colors
    8. Recommendations for balancing the elements
    `;
  } else {
    return `
    请分析以下八字命盘：

    个人信息：
    - 姓名：${baziInput.name || "未知"}
    - 性别：${baziInput.gender === "male" ? "男" : "女"}
    - 出生日期：${baziInput.year}年${baziInput.month}月${baziInput.day}日${baziInput.hour}时${baziInput.minute}分

    八字命盘：
    - 年柱：${baziResult.yearPillar}
    - 月柱：${baziResult.monthPillar}
    - 日柱：${baziResult.dayPillar}
    - 时柱：${baziResult.hourPillar}
    - 日主：${baziResult.dayMaster}
    - 生肖：${baziResult.zodiac}
    - 纳音：${baziResult.nayin}
    - 季节：${baziResult.season}

    五行分布：
    - 木：${baziResult.fiveElements["木"]}
    - 火：${baziResult.fiveElements["火"]}
    - 土：${baziResult.fiveElements["土"]}
    - 金：${baziResult.fiveElements["金"]}
    - 水：${baziResult.fiveElements["水"]}

    ${aspects ? `请重点分析以下方面：${aspects.join("、")}` : "请提供全面分析，包括事业、婚姻、健康和财运等方面。"}

    请提供：
    1. 性格特点分析
    2. 优势与劣势
    3. 事业发展前景
    4. 婚姻情感状况
    5. 健康注意事项
    6. 财运状况
    7. 吉祥时期与颜色
    8. 平衡五行的建议
    `;
  }
}

function getSystemPrompt(language: "zh" | "en"): string {
  if (language === "en") {
    return `
    You are a professional Chinese astrologer and Bazi (Four Pillars of Destiny) expert with deep knowledge of Chinese metaphysics,
    Five Elements theory, and traditional Chinese fortune-telling. Your analysis should be:

    1. Professional and insightful
    2. Balanced and positive, avoiding overly negative predictions
    3. Culturally sensitive and respectful
    4. Practical with actionable advice
    5. Structured with clear sections

    Provide a comprehensive analysis based on the Bazi chart information provided, focusing on giving helpful insights
    rather than deterministic predictions. Emphasize that Bazi is a tool for self-understanding and guidance,
    not absolute fate.
    `;
  } else {
    return `
    你是一位专业的中国占星术大师和八字命理专家，拥有深厚的中国玄学、五行理论和传统算命知识。你的分析应该：

    1. 专业且有洞察力
    2. 平衡积极，避免过于消极的预测
    3. 文化敏感且尊重
    4. 实用且具有可操作性的建议
    5. 结构清晰，分段明确

    基于提供的八字命盘信息进行全面分析，重点提供有用的见解而非决定性预测。强调八字是自我理解和指导的工具，
    而非绝对命运。
    `;
  }
}

function parseAIResponse(
  content: string,
  language: "zh" | "en",
  provider: string = "DeepSeek",
): AIAnalysisResult {
  // 简化处理，实际应用中可能需要更复杂的解析逻辑
  return {
    analysis: content,
    aspects: ["personality", "career", "relationships", "health", "wealth"],
    suggestions:
      language === "en"
        ? [
            "Meditate regularly",
            "Balance your elements",
            "Follow your lucky directions",
          ]
        : ["定期冥想", "平衡五行", "遵循吉祥方向"],
    luckyPeriods:
      language === "en"
        ? ["2025-2026", "2030-2031"]
        : ["2025-2026年", "2030-2031年"],
    warnings:
      language === "en"
        ? [
            "Avoid risky investments in 2024",
            "Pay attention to health in autumn",
          ]
        : ["2024年避免风险投资", "秋季注意健康"],
    provider,
  };
}

File: backend/handlers/ai.go
package handlers

import (
	"bazi_fortune_app/backend/config"
	"bazi_fortune_app/backend/models"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// AIHandler AI处理器
type AIHandler struct {
	config *config.Config
}

// NewAIHandler 创建AI处理器
func NewAIHandler(cfg *config.Config) *AIHandler {
	return &AIHandler{
		config: cfg,
	}
}

// AnalyzeRequest AI分析请求
type AnalyzeRequest struct {
	BaziID   string `json:"bazi_id" binding:"required"`
	Language string `json:"language"` // zh, en
}

// AnalyzeResponse AI分析响应
type AnalyzeResponse struct {
	Success  bool   `json:"success"`
	Message  string `json:"message"`
	Analysis string `json:"analysis,omitempty"`
}

// DeepSeekRequest DeepSeek API请求结构
type DeepSeekRequest struct {
	Model    string    `json:"model"`
	Messages []Message `json:"messages"`
}

// Message DeepSeek消息结构
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// DeepSeekResponse DeepSeek API响应结构
type DeepSeekResponse struct {
	Choices []Choice `json:"choices"`
}

// Choice DeepSeek选择结构
type Choice struct {
	Message Message `json:"message"`
}

// Analyze 使用AI分析八字
func (h *AIHandler) Analyze(c *gin.Context) {
	var req AnalyzeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, AnalyzeResponse{
			Success: false,
			Message: "Invalid request format: " + err.Error(),
		})
		return
	}

	// 解析八字ID
	baziID, err := uuid.Parse(req.BaziID)
	if err != nil {
		c.JSON(http.StatusBadRequest, AnalyzeResponse{
			Success: false,
			Message: "Invalid bazi ID format",
		})
		return
	}

	// 查询八字记录
	var baziData models.BaziData
	if err := models.GetDB().First(&baziData, "id = ?", baziID).Error; err != nil {
		c.JSON(http.StatusNotFound, AnalyzeResponse{
			Success: false,
			Message: "Bazi record not found",
		})
		return
	}

	// 获取八字输入和结果数据
	input, err := baziData.GetInputData()
	if err != nil {
		c.JSON(http.StatusInternalServerError, AnalyzeResponse{
			Success: false,
			Message: "Failed to parse input data",
		})
		return
	}

	result, err := baziData.GetResultData()
	if err != nil {
		c.JSON(http.StatusInternalServerError, AnalyzeResponse{
			Success: false,
			Message: "Failed to parse result data",
		})
		return
	}

	// 设置默认语言
	if req.Language == "" {
		req.Language = "zh"
	}

	// 调用DeepSeek API进行分析
	analysis, err := h.callDeepSeekAPI(input, result, req.Language)
	if err != nil {
		c.JSON(http.StatusInternalServerError, AnalyzeResponse{
			Success: false,
			Message: "Failed to analyze with AI: " + err.Error(),
		})
		return
	}

	// 更新八字记录的分析结果
	baziData.Analysis = &analysis
	if err := models.GetDB().Save(&baziData).Error; err != nil {
		c.JSON(http.StatusInternalServerError, AnalyzeResponse{
			Success: false,
			Message: "Failed to save analysis result",
		})
		return
	}

	c.JSON(http.StatusOK, AnalyzeResponse{
		Success:  true,
		Message:  "AI analysis completed",
		Analysis: analysis,
	})
}

// callDeepSeekAPI 调用DeepSeek API
func (h *AIHandler) callDeepSeekAPI(input *models.BaziInput, result *models.BaziResult, language string) (string, error) {
	// 构建提示词
	prompt := h.buildPrompt(input, result, language)

	// 构建请求
	reqBody := DeepSeekRequest{
		Model: "deepseek-chat",
		Messages: []Message{
			{
				Role:    "system",
				Content: h.getSystemPrompt(language),
			},
			{
				Role:    "user",
				Content: prompt,
			},
		},
	}

	// 序列化请求
	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	// 创建HTTP请求
	url := h.config.Deepseek.APIBaseURL + "/chat/completions"
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	// 设置请求头
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+h.config.Deepseek.APIKey)

	// 发送请求
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	// 读取响应
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("API request failed with status %d: %s", resp.StatusCode, string(body))
	}

	// 解析响应
	var deepSeekResp DeepSeekResponse
	if err := json.Unmarshal(body, &deepSeekResp); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	if len(deepSeekResp.Choices) == 0 {
		return "", fmt.Errorf("no choices in response")
	}

	return deepSeekResp.Choices[0].Message.Content, nil
}

// buildPrompt 构建分析提示词
func (h *AIHandler) buildPrompt(input *models.BaziInput, result *models.BaziResult, language string) string {
	if language == "en" {
		return fmt.Sprintf(`Please analyze the following Bazi (Chinese astrology) chart:

Birth Information:
- Date: %d-%02d-%02d %02d:%02d
- Gender: %s
- Name: %s

Bazi Chart:
- Year Pillar: %s
- Month Pillar: %s  
- Day Pillar: %s
- Hour Pillar: %s
- Day Master: %s
- Zodiac: %s
- Nayin: %s
- Season: %s

Five Elements Distribution:
- Wood: %d, Fire: %d, Earth: %d, Metal: %d, Water: %d

Please provide a comprehensive analysis covering:
1. Personality traits and characteristics
2. Career and wealth prospects
3. Relationships and marriage
4. Health considerations
5. Lucky elements, colors, numbers, and directions
6. Life advice and recommendations

Please write in English and provide detailed explanations.`,
			input.Year, input.Month, input.Day, input.Hour, input.Minute,
			input.Gender, input.Name,
			result.YearPillar, result.MonthPillar, result.DayPillar, result.HourPillar,
			result.DayMaster, result.Zodiac, result.Nayin, result.Season,
			result.FiveElements["木"], result.FiveElements["火"], result.FiveElements["土"],
			result.FiveElements["金"], result.FiveElements["水"])
	}

	return fmt.Sprintf(`请分析以下八字命盘：

出生信息：
- 出生日期：%d年%02d月%02d日 %02d时%02d分
- 性别：%s
- 姓名：%s

八字排盘：
- 年柱：%s
- 月柱：%s
- 日柱：%s
- 时柱：%s
- 日主：%s
- 生肖：%s
- 纳音：%s
- 季节：%s

五行分布：
- 木：%d个，火：%d个，土：%d个，金：%d个，水：%d个

请提供全面的命理分析，包括：
1. 性格特点和个性分析
2. 事业财运发展趋势
3. 感情婚姻状况
4. 健康注意事项
5. 有利的五行、颜色、数字、方位
6. 人生建议和改运方法

请用中文详细分析，字数在3000字以上。`,
		input.Year, input.Month, input.Day, input.Hour, input.Minute,
		input.Gender, input.Name,
		result.YearPillar, result.MonthPillar, result.DayPillar, result.HourPillar,
		result.DayMaster, result.Zodiac, result.Nayin, result.Season,
		result.FiveElements["木"], result.FiveElements["火"], result.FiveElements["土"],
		result.FiveElements["金"], result.FiveElements["水"])
}

// getSystemPrompt 获取系统提示词
func (h *AIHandler) getSystemPrompt(language string) string {
	if language == "en" {
		return `You are a professional Chinese astrology (Bazi) master with deep knowledge of traditional Chinese metaphysics. You specialize in analyzing Bazi charts and providing insightful, accurate, and helpful interpretations. Your analysis should be comprehensive, detailed, and practical, helping people understand their destiny and make better life decisions.`
	}

	return `你是一位专业的八字命理大师，精通中华传统玄学文化，擅长分析八字命盘并提供准确、有用的解读。你的分析应该全面、详细、实用，帮助人们了解自己的命运并做出更好的人生决策。请用专业而通俗易懂的语言进行分析。`
}
